patroni/postgresql.py

```diff
@@ -173,14 +180,32 @@ def pg_ctl(self, cmd, *args, **kwargs):
         pg_ctl = [self._pgcommand('pg_ctl'), cmd]
         if cmd in ('start', 'stop', 'restart'):
             pg_ctl += ['-w']
-            timeout = self.config.get('pg_ctl_timeout')
+            timeout = kwargs.pop('timeout', None)
+            timeout = self.config.get('pg_ctl_timeout', timeout)
```

Comment:
I didn't want to rely on pg_ctl timeout because I still want to allow for manual failover. For example, master_start_timeout is configured to a large number, but sysadmin decides that in this case it's more important to fail over instead. Maybe that usecase could be taken care of by just stopping Patroni. Not sure if it would simplify anything because the current code also handles PostgreSQL restart_after_crash=true behavior (else branch in Postgresql.check_for_startup).

Merging check_for_startup, post_recover and is_healthy could result in some simplification though.

As for stop timing out, this is why demote_asap does stop('immediate', checkpoint=False). That can still fail to stop within a timeout, but there is little we can do from within Patroni except hope that the load balancer health check will mark us unhealthy and take care of redirecting connections. I do have a patch coming up that uses kernel watchdog support to ensure that the node goes away in case things take too long. I made a note to check how other call sites behave in case of stop time outs.